solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java

```diff
@@ -0,0 +1,587 @@
+            }
+          }
+
+          //now go through all other replicas and unset previous leader
+          for (String r : allReplicas) {
+            State st = rs.get(r);
+            if (st == null) continue;//unlikely
+            if (!Objects.equals(r, next)) {
+              if (st.isLeader) {
+                //some other replica is the leader now. unset
+                ops.add(new Op(Op.Type.ADD, new State(st.replica, st.state, Boolean.FALSE, st.version + 1)));
+                ops.add(new Op(Op.Type.DELETE, st));
+              }
+            }
+          }
+          if (log.isDebugEnabled()) {
+            log.debug("flipLeader on:{}, {} -> {}, ops: {}", rs.path, allReplicas, next, ops);
+          }
+          return ops;
+        }
+
+      }.init(rs);
+    }
+
+    /**
+     * Delete a replica entry from per-replica states
+     *
+     * @param replica name of the replica to be deleted
+     */
+    public static WriteOps deleteReplica(String replica, PerReplicaStates rs) {
+      return new WriteOps() {
+        @Override
+        protected List<Op> refresh(PerReplicaStates rs) {
+          List<Op> result;
+          if (rs == null) {
+            result = Collections.emptyList();
+          } else {
+            State state = rs.get(replica);
+            result = addDeleteStaleNodes(new ArrayList<>(), state);
+          }
+          return result;
+        }
+      }.init(rs);
+    }
+
+    public static WriteOps addReplica(String replica, Replica.State state, boolean isLeader, PerReplicaStates rs) {
+      return new WriteOps() {
+        @Override
+        protected List<Op> refresh(PerReplicaStates rs) {
+          return singletonList(new Op(Op.Type.ADD,
+              new State(replica, state, isLeader, 0)));
+        }
+      }.init(rs);
+    }
+
+    /**
+     * mark a bunch of replicas as DOWN
+     */
+    public static WriteOps downReplicas(List<String> replicas, PerReplicaStates rs) {
+      return new WriteOps() {
+        @Override
+        List<Op> refresh(PerReplicaStates rs) {
+          List<Op> ops = new ArrayList<>();
+          for (String replica : replicas) {
+            State r = rs.get(replica);
+            if (r != null) {
+              if (r.state == Replica.State.DOWN && !r.isLeader) continue;
+              ops.add(new Op(Op.Type.ADD, new State(replica, Replica.State.DOWN, Boolean.FALSE, r.version + 1)));
+              addDeleteStaleNodes(ops, r);
+            } else {
+              ops.add(new Op(Op.Type.ADD, new State(replica, Replica.State.DOWN, Boolean.FALSE, 0)));
+            }
+          }
+          if (log.isDebugEnabled()) {
+            log.debug("for coll: {} down replicas {}, ops {}", rs, replicas, ops);
+          }
+          return ops;
+        }
+      }.init(rs);
+    }
+
+    /**
+     * Just creates and deletes a summy entry so that the {@link Stat#getCversion()} of states.json
+     * is updated
+     */
+    public static WriteOps touchChildren() {
```

Comment:
Isn't reading the `state.json` znode Stat sufficient?
If it is not and this can't be simplified, then doesn't it break the logic used elsewhere of comparing the stored `cversion` to the actual znode `cversion` to decide if the children should be read again?

Reply:
This is a rare and we always ensure that if `cversion` is same the content is same